<div style="border:solid Chocolate 2px; padding: 40px">

<b>Привет!👋</b>

Меня зовут Огурлиев Анзор, я буду ревьюером твоего проекта. Сразу хочу предложить тебе общаться на "ты", если тебе это неудобно, скажи мне об этом. Нам предстоит пройти с тобой через тернистый и от того интересный путь анализа данных! Можешь не переживать, я всегда помогу тебе, если у тебя возникнут вопросы)

По ходу твоей работы я буду оставлять свои комментарии, по выполненным тобой шагам. Попрошу тебя не удалять их, так мне будет проще отслеживать твой прогресс в случае повторной проверки. 
    
Очень важно вести диалог и обсуждать те пункты, которые у тебя могли не получиться, поэтому, ты можешь оставлять свои комментарии, и мы обязательно во всем разберемся! Самое главное - выделяй свои комментарии удобным для тебя способом (цвет, рамка), но так, чтобы они отчетлитво отличались от моих комментариев. Так мы ни в чем не запутаемся)

Свои комментарии я буду выделять так: <font color='green'>зеленый</font>, <font color='gold'>желтый</font> и <font color='red'>красный</font> цвет, например:

<br/>

<div class="alert alert-success">
<h2> Комментарий ревьюера V1<a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Этот комментарий я буду оставлять, если твое решение оптимально для данного шага.
</div>

<br/>

<div class="alert alert-warning">
    <h2> Комментарий ревьюера V1<a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> Это не означает, что ты совершил ошибку. Такими комментариями я буду отмечать места, где ты справился с задачей, но её можно решить ещё лучше. Также в таких комментариях я буду оставлять для тебя другие советы.
</div>


<br/>
<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера V1<a class="tocSkip"></h2>

    
<b>На доработку🤔:</b>
Если решение, которое ты применил на определенном шаге, требует значительных изменений, то я буду оставлять такой комментарий. Тебе нужно будет подумать еще раз как правильным образом решить данный этап. Если ревью содержит такие комментарии, то я не смогу принять проект с первого раза.
</div>
    
P.S. В оглавлении комментария ты увидишь `V1` (`V1`, `V2`, `V3`) - это версия моего комментария, которая понадобится нам с тобой, чтобы не запутаться, если потребуется оставить тебе комментарии после первой проверки.

# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

In [5]:
data_null = data[::]

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [6]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [7]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [8]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [9]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [10]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [11]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [12]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [13]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [14]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [15]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [16]:
data.duplicated().sum()

71

In [17]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [18]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [19]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [20]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [21]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [22]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

Для ответа на поставленный вопрос необходимо подсчитать долю тех заемщиков, у которых были задолженности по возврату кредита (значение 1 в столбце debt) по отношению к общему количеству кредитов, выданных клиентам категории "количество детей".
Это можно увидеть в сводной таблице. В этой таблице группировка будет по столбцу *children*, столбец из которого будем брать данные для группировки *debt*, а подсчет количества значений будет осуществляться из копии столбца *debt*.

In [23]:
import math
def pirs_chupr(df,lst):  
    '''
    функция для расчета коэффициентов взаимной сопряженности Пирсона и Чупрунова
    Используются для определения тесноты связи, когда качественные признаки состоят более, чем из двух групп
    Параметры функции:
    df - сводная таблица, содержащая сгруппированные данные для расчета
    lst - список уникальных значений, являющихся заголовками столбцов в сводной таблице    
    Основано на материале "Практикум по теории статистики" Учебное пособие под ред. Р.А. Шмоловой
    '''
    #print(lst)
    #Разбор сводной таблицы на список списков
    d = []
    for i in range(len(lst)):
        temp = []
        temp.append(df[lst[i]][0])
        temp.append(df[lst[i]][1])
        temp.append(df[lst[i]].sum())
        d.append(temp)
    col_itog = []
    #print(d)
    #Расчет коэффиентов
    k = len(d[0])
    #print(k)
    for i in range(k-1):
        s = 0
        for j in range(len(d)):
            s+=d[j][i]
        col_itog.append(s)
    
    sum_divis = []
    for j in range(k-1):
        su = 0
        
        for i in range(len(d)):
            su += (d[i][j]**2)/d[i][2]
        sum_divis.append(su/col_itog[j])   
    #print(col_itog)
    #print(sum_divis)
    #Коэффициент Пирсона
    kp = sum(sum_divis) - 1
    #Коэффициент Чупрунова
    kch = math.sqrt(kp/((len(d) - 1)))
    return(kp, kch)

<div class="alert alert-warning">
<h2> Комментарий ревьюера V1<a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> Подскажи, пожалуйста, ты сам составил данную функцию или нашел её в интернете?

<div class="alert alert-info">
<h2> Ответ <a class="tocSkip"> </h2>
    
Функция написана самостоятельно


In [24]:
#получаем копию столбца debt.
data['debt1'] = data['debt']
#формирование сводной таблицы
data_pivot = data.pivot_table(index=['children'], columns='debt', values='debt1', aggfunc='count')

In [25]:
#расчет доли имевщих проблемы с выплатой кредита к общему числу заемщиков категории "количество детей"
#0 - кол-во заемщиков без проблем возврата, 1 - кол-во заемщиков с долгами, perc_debt - доля должников
data_pivot['perc_debt'] = data_pivot[1]/(data_pivot[0]+data_pivot[1])
data_pivot = data_pivot.fillna(0)
data_pivot.sort_values(by='perc_debt')

debt            0       1  perc_debt
children                            
5             9.0     0.0   0.000000
0         13028.0  1063.0   0.075438
3           303.0    27.0   0.081818
1          4364.0   444.0   0.092346
2          1858.0   194.0   0.094542
4            37.0     4.0   0.097561

<div class="alert alert-success">
<h2> Комментарий ревьюера V1<a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Правильно построил сводную таблицу, молодец!
</div>

**Вывод:** 
Из сводной таблицы видно, что доля должников колеблется в пределах 2% независимо от количества детей. Кроме случая с количеством 5 детей. Но в данном случае количество заемщиков слишком мало, чтобы опираться в суждениях на это.
Если для кредитной организации 2% - это достаточный показатель, то тогда зависимость прослеживается, но в общем случае,  **зависимости между количеством детей и возвратом кредита в срок НЕТ**

Значения коэффициентов очень далеки от 1, поэтому связь практически отсутствует 


<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера V1<a class="tocSkip"></h2>

    
<b>На доработку🤔:</b>
Прежде хотел бы услышать от тебя почему ты используешь данный коэффициент для выявления силы зависимости. Можно ли этот коэффициент сопряженности использовать в данном случае, подходят ли наши признаки для этого? Также нужно помнить, что никак не обработаны выбросы. 
    
Разница в несколько процентов может быть значимой для нашего исследования, также нужно учитывать количество клиентов в разных группах. Можно ли учитывать все группы для анализа? 
</div>

<div class="alert alert-info">
<h2> Ответ <a class="tocSkip"> </h2>
    
Коэффициент Пирсона-Чупрова был взят по причине отсутствия критерия для сравнения от заказчика. Не было понятно, какая разница между долями должников является существенной чтобы делать вывод о зависимости. Кроме того, эти коэффициенты используются для выявления тесноты связи между атрибутивными признаками. Теперь понятно, что несколько процентов - это существенно, а "несколько" - это 2 и более.

<div class="alert alert-success">
<h2> Комментарий ревьюера V2<a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> На самом деле на реальной работе мало когда заказчик придет с строгими критериями. Тебе дадут общую задачу, а критерии с помощью которых ты будешь проводить анализ уже на твое усмотрение. Заказчик же не знает какие методы используются для определенных данных.
</div>

Из сводной таблицы видно, что данные о заемах "многодетных" семей (3, 4 и 5 детей) можно отбросить и не использовать в анализе по причине их незначительного количества, а соответственно слабого влияния на итоговый результат.
Тогда из оставшихся строк можно заключить, что на возврат кредита в срок оказывает влияние *не количество*, а *наличие* детей. Разница в доле невозврата в семьях с одним и двумя детьми около 0.2%. Даже если рассмотреть отброшенные результаты с 3 детьми, то разница составит чуть более 1%. 

**Вывод  (новый):**
**Количество детей** на возврат кредита в срок **не влияет**. Влияет их наличие

<div class="alert alert-success">
<h2> Комментарий ревьюера V2<a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Именно так, все правильно.
</div>

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [26]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.
#получаем копию столбца debt.
data['debt1'] = data['debt']
#формирование сводной таблицы
data_pivot = data.pivot_table(index=['family_status'], columns='debt', values='debt1', aggfunc='count')
#расчет доли имевщих проблемы с выплатой кредита к общему числу заемщиков категории "количество детей"
data_pivot['perc_debt'] = data_pivot[1]/(data_pivot[0]+data_pivot[1])
data_pivot = data_pivot.fillna(0)

#data_pivot = data.pivot_table(index=['debt'], columns='children', values='debt1', aggfunc='count')

In [27]:
data_pivot.sort_values(by='perc_debt')

debt                       0    1  perc_debt
family_status                               
вдовец / вдова           888   63   0.066246
в разводе               1105   84   0.070648
женат / замужем        11334  927   0.075606
гражданский брак        3749  385   0.093130
Не женат / не замужем   2523  273   0.097639

<div class="alert alert-success">
<h2> Комментарий ревьюера V1<a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Видно, что ты хорошо справляешься с методами построения сводных таблиц, так держать!
</div>

**Вывод (старый и неправильный):** 

Зависимости между семейным положением и возвратом кредита в срок НЕТ

<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера V1<a class="tocSkip"></h2>

    
<b>На доработку🤔:</b>
Зависимость от данного признака есть. Попробуй выявить зависимость из сводной таблицы. В данном случае коэффициенты корреляции не отражают истинные взаимосвязи. Коэффициент показывает, что признаки в какой-то степени зависят друг от друга, но строгой зависимости коэффициент не отражает. Также стоит помнить, что корреляция может быть нелинейной или может присутствовать скрытое влияние от третей переменной.
</div>

<div class="alert alert-info">
<h2> Доработка <a class="tocSkip"> </h2>


**Вывод (новый)** 

Если сделать предположение, что заемщики, состоящие в гражданском браке, ранее не были в официальном браке, то на основании данных из сводной таблицы можно заключить, что опыт ОФИЦИАЛЬНОЙ семейной жизни добавляет отвественности. Таким образом, **зависимость** между семейным положением и возвратом кредита в срок **имеется**. 

<div class="alert alert-success">
<h2> Комментарий ревьюера V2<a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Абсолютно верно, клиенты, побывавшие в браке, реже допускают просрочки.
</div>

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [28]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.
# Ваш код будет здесь. Вы можете создавать новые ячейки.
#получаем копию столбца debt.
data['debt1'] = data['debt']
#формирование сводной таблицы
data_pivot = data.pivot_table(index=['total_income_category'], columns='debt', values='debt1', aggfunc='count')
#расчет доли имевщих проблемы с выплатой кредита к общему числу заемщиков категории "количество детей"
data_pivot['perc_debt'] = data_pivot[1]/(data_pivot[0]+data_pivot[1])
data_pivot = data_pivot.fillna(0)

In [29]:
data_pivot.sort_values(by='perc_debt')

debt                       0     1  perc_debt
total_income_category                        
D                        328    21   0.060172
B                       4660   354   0.070602
A                         23     2   0.080000
C                      14568  1353   0.084982
E                         20     2   0.090909

**Вывод (старый и неправильный):** Данные в столбце **perc_debt** сводной таблицы свидетельствуют о том, что среди всех категорий благосостояния процент неплательщиков примерно одинаков. Коэффициенты Пирсона и Чупрунова подтверждают, что **зависимости между уровнем дохода и возвратом кредита в срок НЕТ**

<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера V1<a class="tocSkip"></h2>

    
<b>На доработку🤔:</b>
Нужно выявлять зависимость по сводной таблице. Коэффициент корреляции не дает ответа на поставленный вопрос.    

Обрати внимание на распределение клиентов по группам. Можно ли учитывать все группы при выявлении зависимости?
    
Подсказка: зависимость нужно определять только по 2 группам с достаточным количеством клиентов.
</div>

<div class="alert alert-info">
<h2> Доработка <a class="tocSkip"> </h2>

Данные, приведенные в сводной таблице свидетельствуют о том, что анализ можно проводить только по двум группам данных из пяти, что вряд ли может привести к правильным умозаключениям. Тем не менее, даже эти две группы показывают, что разница между долями невозврата составляет лишь 1,5%. 

**Вывод (новый)** 
**Зависимость** между уровнем дохода и возвратом кредита в срок если и есть, то **слабая**

<div class="alert alert-success">
<h2> Комментарий ревьюера V2<a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Эти две группы составляют 98% от всей выборки, это значит, что они описывают полностью тенденции от дохода. Если бы мы рассматривали 50% всех данных, то да, можно сомневаться, что выявленные зависимости будут аналогичными для других клиентов.
</div>

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [30]:
#получаем копию столбца debt.
data['debt1'] = data['debt']
#формирование сводной таблицы
data_pivot = data.pivot_table(index=['purpose_category'], columns='debt', values='debt1', aggfunc='count')
#расчет доли имевщих проблемы с выплатой кредита к общему числу заемщиков категории "количество детей"
data_pivot['perc_debt'] = data_pivot[1]/(data_pivot[0]+data_pivot[1])
data_pivot = data_pivot.fillna(0)

In [31]:
data_pivot.sort_values(by='perc_debt')

debt                         0    1  perc_debt
purpose_category                              
операции с недвижимостью  9971  780   0.072551
проведение свадьбы        2130  183   0.079118
получение образования     3619  369   0.092528
операции с автомобилем    3879  400   0.093480

**Вывод: (старый и неправильный)** 

Данные в столбце **perc_debt** сводной таблицы свидетельствуют о том, что на какие бы цели люди не брали кредит, доля тех, кто никуда не торопится практически одинакова. Коэффициенты Пирсона и Чупрунова подтверждают, что **цели кредита НИКАК не влияют на возврат его в срок**

<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера V1<a class="tocSkip"></h2>

    
<b>На доработку🤔:</b>
Аналогичный комментарий, коэффициенты корреляции не дают ответ на поставленный вопрос, попробуй выявить зависимость по сводной таблице.
</div>

<div class="alert alert-info">
<h2> Доработка <a class="tocSkip"> </h2>

Процент невозврата кредита в срок у заемщиков, проводящих операции с недвижимостью ниже, чем у тех, кто связал свою жизнь в данный момент с автомобилем или получением образования. Возвожно, это связано с более низким процентом по ипотеке или более суровыми требованиями к данному виду кредитования. Возврат по свадьбам тоже более активный, видимо, суммы поменьше.

**Вывод (новый)**

Цели кредита **влияют** на его возврат в срок.

<div class="alert alert-success">
<h2> Комментарий ревьюера V2<a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Все правильно указал, продолжай в том же духе!
</div>

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

В связи с тем, что данные пропущены только столбцах **days_employed** и **total_income** и количество пропущенных в столбцах точно совпадает, то есть гипотеза, что пропуск данных связан с машинной обработкой этих данных после введения.
Для этого можно, например, найти долю отсутствующих данных в разрезе какого-либо столбца с категориальными данными и если она будет примерно одинаковой у всех, то вряд ли это делали специально в **ручном** режиме.
Так как с пропусками боролись, основываясь на группировке по типу занятости, то рассчитаем долю отсутствующих данных в разрезе типа занятости

In [32]:
col = 'income_type'

In [33]:
income_type_unique = data_null[col].unique()

In [34]:
print('Доля отсутствующих данных в разрезе типа занятости')
for i in income_type_unique:
    a = data_null.loc[data_null[col] == i,'days_employed'].isna().sum()/data_null.loc[data_null[col] == i,'children'].isna().count()
    print(f'{i} - {a:.2f}')
print('----------------------------')
print('Общее количество заемщиков в разрезе типа занятости')
for i in income_type_unique:
    print(i , data_null.loc[data_null[col] == i,'children'].isna().count())

Доля отсутствующих данных в разрезе типа занятости
сотрудник - 0.10
пенсионер - 0.11
компаньон - 0.10
госслужащий - 0.10
безработный - 0.00
предприниматель - 0.50
студент - 0.00
в декрете - 0.00
----------------------------
Общее количество заемщиков в разрезе типа занятости
сотрудник 11119
пенсионер 3856
компаньон 5085
госслужащий 1459
безработный 2
предприниматель 2
студент 1
в декрете 1


Из расчета видно, что доля отсутствующих данных в разрезе типа занятости практически одинакова во всех категориях работников, чья численность является значимой. Это массовая ошибка. Вряд ли банк выдал заем при отсутствии введенных данных в этих столбцах. Значит они были и пропали. Маловероятно, что служащие сидели и удаляли данные клавишей Delete. Отрицательные значения в **days_employed** косвенно подтверждают ошибку в формулах для расчета. 
Доля в 10% свидельствует о том, что ошибку вовремя заметили и предотвратили потерю всех данных.

*Ответ:* Причиной появления пропусков в исходных данных является некорректная машинная обработка

<div class="alert alert-success">
<h2> Комментарий ревьюера V1<a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Все правильно, молодец!
</div>

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* Медианное значение является лучшим решением для заполнения отсуствующих значений количественных переменных по причине того, что они находятся посередине выборки и являются истиным средним значением, в то время как среднее значение может иметь большие отклонения.

<div class="alert alert-success">
<h2> Комментарий ревьюера V1<a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Абсолютно верно.
</div>

### Шаг 4: общий вывод.

В результате исследования не было обнаружено связей между указанными категориями и возвратом кредита в срок. Единственная категория, расчеты по которой могут быть подвергнуты сомнению это "уровнень дохода", по причине отсутствия реальных исходных данных в объеме около 10% в столбце **total_income**, на основе которых и создавалась эта категория. 
**Рекомендация** необходимо провести исследование на основе совокупности факторов.

<div class="alert alert-warning">
<h2> Комментарий ревьюера V1<a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> В итоговом выводе нужно достаточно подробно написать основные пункты выявленных зависимостей. Стоит доработать вывод с учетом моих комментариев по ходу работы.

<div class="alert alert-info">
<h2> Доработка <a class="tocSkip"> </h2>

**Шаг 4: общий вывод**

В ходе исследования были проведены расчеты для выявления зависимости возврата кредита в срок от нескольких показателей:
- от количества детей
- от семейного положения заемщиков
- от уровня дохода
- от целей кредита

Проведенное исследование показало, что:

1. Многодетные семьи (3 ребенка и более) берут не так много кредитов, чтобы их показатели можно было использовать в анализе ситуации. Из оставшихся категорий распределение следующее:
| кол-во детей   | Доля должников|
| ------------- |:-------------:|
| 0             | 7.5% |
| 1             | 9.2% |
| 2 | 9.5%      |

Поэтому именно количество детей не влияет на возврат кредита в срок, а вот их наличие - влияет

----

2. При выявлении зависимости возврата кредита в срок от семейного положения оказалось, что заемщики, имеющие отношение к официальному браку, более ответственно относятся к возврату кредита. То есть зависимость есть
| Сем. полож   | Доля должников| Брак|
| ------------- |:-------------:|----|
| вдовец / вдова | 6.6% |+
| в разводе      | 7.1% |+
| женат / замужем | 7.5% |+
|гражданский брак|9.3%|-
|не женат/не замужем|9.8%|-

----

3. Выявлена зависимость возврата кредита в срок от дохода заемщиков. Но, по причине того, что из пяти категорий для анализа подходит только две и разница между долями должников составила всего 1.5%, судить о тесной зависимости сложно.
| Доход   | Доля должников|
| ------------- |:-------------:|
| 200001–1000000| 7% |
| 50001–200000 | 8.5% |

----

4. Выявлена связь между сроком возврата и целью кредита. Граждане, берущие кредиты на недвижимость, погашают их более дисциплинировано, чем, например, те, кто решил обзавестись автомобилем. 
| Доход   | Доля должников|
| ------------- |:-------------:|
|операции с недвижимостью|7.2%|
|проведение свадьбы|7.9%|
|получение образования|9.2%|
|операции с автомобилем|9.3%|	

___

В двух исследованиях из четырех пришлось отбрасывать группы для анализа, так как в них было мало данных. Поэтому, скорее всего, необходимо менять критерии для группировки или по другому формулировать задачу ислледования.
Также было замечено, что во всех исследованиях результаты (доля недобросовестных заещиков) расположены около значения 8%. Минимальный - 6.6%, максимальный 9.6%. Это дает основание предполагать, что существуют дополнительные факторы, оказывающие влияние на конечный результат.

**Рекомендация** необходимо провести исследование на основе совокупности факторов.

<div class="alert alert-success">
<h2> Комментарий ревьюера V2<a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Отличный вывод, все четко, понятно и структурировано. Вывод в полной мере отражает существующие зависимости, и по нему можно принимать дальнейшие решения. Молодец, что оставляешь свои рекомендации.
</div>

<div style="border:solid Chocolate 2px; padding: 40px">

**Общий вывод по проекту**

Спасибо за твой проект! Я уверен, что ты приложил все усилия, чтобы сделать этот проект. Это видно по твоей работе, а сформулированные выводы отражают, насколько успешно ты осваиваешь материал. 

**Отмечу отдельные положительные моменты проекта🙂**:
 
* Ты полностью следовал всем этапам выполнения задания и справился почти со всем. Хорошая работа!
* Отдельно выделю некоторые выводы, которые ты смог сформулировать абсолютно точно!
* У тебя опрятный и читаемый код. Возможно это звучит банально, однако при работе с чужим кодом это крайне важно)


**На что стоит обратить внимание🤓**:
    
* Важно доработать итоговый вывод.
* Попробуй заново выявить зависимости в каждом пункте.
* Остальные рекомендации ты найдешь в моих комментариях!


**Желаю удачи и жду твой проект на повторное ревью!**😉

<div style="border:solid Chocolate 2px; padding: 40px">

Ты отлично справился с данной работой! Видно, как ты успешно осваиваешь материал, применяешь полученные знания на практике. Мне было очень приятно проверять твою работу, надеюсь, мои комментарии и советы были для тебя полезными.

**Желаю удачи в последующем обучении и исследованиях!**😉